In [24]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

In [25]:
df = pd.read_csv("2017-fordgobike-tripdata.csv")

In [3]:
df

duration_sec                start_time                  end_time  \
0              80110  2017-12-31 16:57:39.6540  2018-01-01 15:12:50.2450   
1              78800  2017-12-31 15:56:34.8420  2018-01-01 13:49:55.6170   
2              45768  2017-12-31 22:45:48.4110  2018-01-01 11:28:36.8830   
3              62172  2017-12-31 17:31:10.6360  2018-01-01 10:47:23.5310   
4              43603  2017-12-31 14:23:14.0010  2018-01-01 02:29:57.5710   
...              ...                       ...                       ...   
519695           435  2017-06-28 10:00:54.5280  2017-06-28 10:08:10.4380   
519696           431  2017-06-28 09:56:39.6310  2017-06-28 10:03:51.0900   
519697           424  2017-06-28 09:47:36.3470  2017-06-28 09:54:41.1870   
519698           366  2017-06-28 09:47:41.6640  2017-06-28 09:53:47.7150   
519699           188  2017-06-28 09:49:46.3770  2017-06-28 09:52:55.3380   

        start_station_id                                 start_station_name  \
0                     74                              Laguna St at Hayes St   
1                    284  Yerba Buena Center for the Arts (Howard St at ...   
2                    245                             Downtown Berkeley BART   
3                     60                               8th St at Ringold St   
4                    239                      Bancroft Way at Telegraph Ave   
...                  ...                                                ...   
519695                81                                 Berry St at 4th St   
519696                66                              3rd St at Townsend St   
519697                21   Montgomery St BART Station (Market St at 2nd St)   
519698                58                               Market St at 10th St   
519699                25                                Howard St at 2nd St   

        start_station_latitude  start_station_longitude  end_station_id  \
0                    37.776435              -122.426244              43   
1                    37.784872              -122.400876              96   
2                    37.870348              -122.267764             245   
3                    37.774520              -122.409449               5   
4                    37.868813              -122.258764             247   
...                        ...                      ...             ...   
519695               37.775880              -122.393170              45   
519696               37.778742              -122.392741             321   
519697               37.789625              -122.400811              48   
519698               37.776619              -122.417385              59   
519699               37.787522              -122.397405              48   

                                         end_station_name  \
0       San Francisco Public Library (Grove St at Hyde...   
1                                   Dolores St at 15th St   
2                                  Downtown Berkeley BART   
3            Powell St BART Station (Market St at 5th St)   
4                               Fulton St at Bancroft Way   
...                                                   ...   
519695                                5th St at Howard St   
519696                                      5th at Folsom   
519697                                2nd St at S Park St   
519698                        S Van Ness Ave at Market St   
519699                                2nd St at S Park St   

        end_station_latitude  end_station_longitude  bike_id   user_type  
0                  37.778768            -122.415929       96    Customer  
1                  37.766210            -122.426614       88    Customer  
2                  37.870348            -122.267764     1094    Customer  
3                  37.783899            -122.408445     2831    Customer  
4                  37.867789            -122.265896     3167  Subscriber  
...                      ...                    ...      ...  

# The dataset includes:
an ID number for each bike

how long it was rented for, in seconds

the starting station and end station id's, latitude, longitude of the starting and end stations and starting and end station names

the start and end time

user type, means whether customer or subscriber

There are 519700 rows and 13 columns

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 519700 entries, 0 to 519699
Data columns (total 13 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   duration_sec             519700 non-null  int64  
 1   start_time               519700 non-null  object 
 2   end_time                 519700 non-null  object 
 3   start_station_id         519700 non-null  int64  
 4   start_station_name       519700 non-null  object 
 5   start_station_latitude   519700 non-null  float64
 6   start_station_longitude  519700 non-null  float64
 7   end_station_id           519700 non-null  int64  
 8   end_station_name         519700 non-null  object 
 9   end_station_latitude     519700 non-null  float64
 10  end_station_longitude    519700 non-null  float64
 11  bike_id                  519700 non-null  int64  
 12  user_type                519700 non-null  object 
dtypes: float64(4), int64(4), object(5)
memory usage: 51.5+ MB


In [27]:
df.shape

(519700, 13)

In [28]:
df.columns

Index(['duration_sec', 'start_time', 'end_time', 'start_station_id',
       'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'end_station_id', 'end_station_name',
       'end_station_latitude', 'end_station_longitude', 'bike_id',
       'user_type'],
      dtype='object')

In [29]:
df.dtypes

duration_sec                 int64
start_time                  object
end_time                    object
start_station_id             int64
start_station_name          object
start_station_latitude     float64
start_station_longitude    float64
end_station_id               int64
end_station_name            object
end_station_latitude       float64
end_station_longitude      float64
bike_id                      int64
user_type                   object
dtype: object

In [30]:
df.user_type.unique()

array(['Customer', 'Subscriber'], dtype=object)

## 

In [8]:
df.user_type.value_counts()

Subscriber    409230
Customer      110470
Name: user_type, dtype: int64

In [9]:
df.columns

Index(['duration_sec', 'start_time', 'end_time', 'start_station_id',
       'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'end_station_id', 'end_station_name',
       'end_station_latitude', 'end_station_longitude', 'bike_id',
       'user_type'],
      dtype='object')

In [31]:
df1=df.copy()

In [11]:
df1

duration_sec                start_time                  end_time  \
0              80110  2017-12-31 16:57:39.6540  2018-01-01 15:12:50.2450   
1              78800  2017-12-31 15:56:34.8420  2018-01-01 13:49:55.6170   
2              45768  2017-12-31 22:45:48.4110  2018-01-01 11:28:36.8830   
3              62172  2017-12-31 17:31:10.6360  2018-01-01 10:47:23.5310   
4              43603  2017-12-31 14:23:14.0010  2018-01-01 02:29:57.5710   
...              ...                       ...                       ...   
519695           435  2017-06-28 10:00:54.5280  2017-06-28 10:08:10.4380   
519696           431  2017-06-28 09:56:39.6310  2017-06-28 10:03:51.0900   
519697           424  2017-06-28 09:47:36.3470  2017-06-28 09:54:41.1870   
519698           366  2017-06-28 09:47:41.6640  2017-06-28 09:53:47.7150   
519699           188  2017-06-28 09:49:46.3770  2017-06-28 09:52:55.3380   

        start_station_id                                 start_station_name  \
0                     74                              Laguna St at Hayes St   
1                    284  Yerba Buena Center for the Arts (Howard St at ...   
2                    245                             Downtown Berkeley BART   
3                     60                               8th St at Ringold St   
4                    239                      Bancroft Way at Telegraph Ave   
...                  ...                                                ...   
519695                81                                 Berry St at 4th St   
519696                66                              3rd St at Townsend St   
519697                21   Montgomery St BART Station (Market St at 2nd St)   
519698                58                               Market St at 10th St   
519699                25                                Howard St at 2nd St   

        start_station_latitude  start_station_longitude  end_station_id  \
0                    37.776435              -122.426244              43   
1                    37.784872              -122.400876              96   
2                    37.870348              -122.267764             245   
3                    37.774520              -122.409449               5   
4                    37.868813              -122.258764             247   
...                        ...                      ...             ...   
519695               37.775880              -122.393170              45   
519696               37.778742              -122.392741             321   
519697               37.789625              -122.400811              48   
519698               37.776619              -122.417385              59   
519699               37.787522              -122.397405              48   

                                         end_station_name  \
0       San Francisco Public Library (Grove St at Hyde...   
1                                   Dolores St at 15th St   
2                                  Downtown Berkeley BART   
3            Powell St BART Station (Market St at 5th St)   
4                               Fulton St at Bancroft Way   
...                                                   ...   
519695                                5th St at Howard St   
519696                                      5th at Folsom   
519697                                2nd St at S Park St   
519698                        S Van Ness Ave at Market St   
519699                                2nd St at S Park St   

        end_station_latitude  end_station_longitude  bike_id   user_type  
0                  37.778768            -122.415929       96    Customer  
1                  37.766210            -122.426614       88    Customer  
2                  37.870348            -122.267764     1094    Customer  
3                  37.783899            -122.408445     2831    Customer  
4                  37.867789            -122.265896     3167  Subscriber  
...                      ...                    ...      ...  

### Top 10 bikes details which are highest duration ?

In [32]:
df1.sort_values('duration_sec',ascending=False).head(10)

duration_sec                start_time                  end_time  \
138862         86369  2017-11-12 10:44:32.0430  2017-11-13 10:44:01.8200   
223508         86355  2017-10-19 11:24:01.8070  2017-10-20 11:23:17.7370   
54888          86325  2017-12-09 10:46:05.9190  2017-12-10 10:44:51.1130   
54889          86281  2017-12-09 10:46:42.2310  2017-12-10 10:44:44.1610   
120423         86252  2017-11-18 12:16:33.2300  2017-11-19 12:14:05.2440   
78414          86228  2017-12-03 00:43:48.0190  2017-12-04 00:40:56.7670   
427904         86199  2017-08-17 18:37:26.1850  2017-08-18 18:34:05.3890   
404193         86135  2017-08-26 16:19:00.8040  2017-08-27 16:14:36.2080   
392030         86075  2017-08-30 11:59:34.8420  2017-08-31 11:54:10.3870   
54034          85975  2017-12-09 16:03:08.5210  2017-12-10 15:56:03.9020   

        start_station_id             start_station_name  \
138862               308               San Pedro Square   
223508               123           Folsom St at 19th St   
54888                212                  Mosswood Park   
54889                212                  Mosswood Park   
120423               245         Downtown Berkeley BART   
78414                211            Broadway at 40th St   
427904               197    El Embarcadero at Grand Ave   
404193               323             Broadway at Kearny   
392030                61            Howard St at 8th St   
54034                  6  The Embarcadero at Sansome St   

        start_station_latitude  start_station_longitude  end_station_id  \
138862               37.336802              -121.894090             308   
223508               37.760594              -122.414817              70   
54888                37.824931              -122.260479             212   
54889                37.824931              -122.260479             212   
120423               37.870348              -122.267764              18   
78414                37.827757              -122.256716             212   
427904               37.808848              -122.249680             233   
404193               37.798014              -122.405950               9   
392030               37.776513              -122.411306             122   
54034                37.804770              -122.403234             324   

                           end_station_name  end_station_latitude  \
138862                     San Pedro Square             37.336802   
223508               Central Ave at Fell St             37.773311   
54888                         Mosswood Park             37.824931   
54889                         Mosswood Park             37.824931   
120423        Telegraph Ave at Alcatraz Ave             37.850222   
78414                         Mosswood Park             37.824931   
427904                   12th St at 4th Ave             37.795812   
404193               Broadway at Battery St             37.798572   
392030                19th St at Mission St             37.760299   
54034   Union Square (Powell St at Post St)             37.788300   

        end_station_longitude  bike_id   user_type  
138862            -121.894090     2231    Customer  
223508            -122.444293      229    Customer  
54888             -122.260479     3482    Customer  
54889             -122.260479     3418    Customer  
120423            -122.260172     1214    Customer  
78414             -122.260479     2673    Customer  
427904            -122.255555     1291    Customer  
404193            -122.400869     2509    Customer  
392030            -122.418892       75  Subscriber  
54034             -122.408531     2143    Customer

### Total number of bikes used in the dataset?

In [33]:
len(df1["bike_id"].unique())

3673

### which bike is used for more number of times ?

In [34]:
df1["bike_id"].value_counts()

68      457
2178    426
210     408
813     403
602     402
       ... 
3623      1
396       1
3618      1
3323      1
1447      1
Name: bike_id, Length: 3673, dtype: int64

bike with id "68" used highest time compared to all the bikes, and it is used for 457

### What is the total number of Customer type users?

In [35]:
print("Total number of Customer type users : ",len(df1[df1["user_type"]=="Customer"]))

Total number of Customer type users :  110470


There are 110470 Customer type users

### What is the total number of Customer type users ?

In [36]:
print("Total number of Subscriber type users : ",len(df1[df1["user_type"]=="Subscriber"]))

Total number of Subscriber type users :  409230


There are 409230 Subscriber type users

### Bikes list which have starting and end stations as same?

In [37]:
starting_stations=set(df1["start_station_name"])
ending_stations=set(df1["end_station_name"])

In [38]:
for i in starting_stations:
    d=df1[((df1["start_station_name"]==i) & (df1["start_station_name"]==i))]

duration_sec                start_time                  end_time  \
1308             140  2017-12-31 09:45:30.2010  2017-12-31 09:47:50.4040   
2915            1006  2017-12-29 21:45:41.2250  2017-12-29 22:02:27.6540   
2917            1015  2017-12-29 21:45:11.1610  2017-12-29 22:02:06.3310   
3006             664  2017-12-29 19:45:00.8570  2017-12-29 19:56:05.6680   
3563            1401  2017-12-29 15:09:46.6360  2017-12-29 15:33:07.8160   
...              ...                       ...                       ...   
505890          2351  2017-07-12 20:59:56.7640  2017-07-12 21:39:08.4470   
506183          1279  2017-07-12 17:52:44.1640  2017-07-12 18:14:03.2870   
507540           403  2017-07-11 19:33:26.7920  2017-07-11 19:40:10.6640   
507688          8210  2017-07-11 16:10:08.2800  2017-07-11 18:26:58.4900   
508021          2031  2017-07-11 16:14:52.1460  2017-07-11 16:48:44.1150   

        start_station_id         start_station_name  start_station_latitude  \
1308                 247  Fulton St at Bancroft Way               37.867789   
2915                 247  Fulton St at Bancroft Way               37.867789   
2917                 247  Fulton St at Bancroft Way               37.867789   
3006                 247  Fulton St at Bancroft Way               37.867789   
3563                 247  Fulton St at Bancroft Way               37.867789   
...                  ...                        ...                     ...   
505890               247  Fulton St at Bancroft Way               37.867789   
506183               247  Fulton St at Bancroft Way               37.867789   
507540               247  Fulton St at Bancroft Way               37.867789   
507688               247  Fulton St at Bancroft Way               37.867789   
508021               247  Fulton St at Bancroft Way               37.867789   

        start_station_longitude  end_station_id  \
1308                -122.265896             245   
2915                -122.265896             170   
2917                -122.265896             170   
3006                -122.265896             267   
3563                -122.265896             240   
...                         ...             ...   
505890              -122.265896             247   
506183              -122.265896             192   
507540              -122.265896             245   
507688              -122.265896             247   
508021              -122.265896             177   

                       end_station_name  end_station_latitude  \
1308             Downtown Berkeley BART             37.870348   
2915           Telegraph Ave at 58th St             37.844493   
2917           Telegraph Ave at 58th St             37.844493   
3006            Derby St at College Ave             37.861804   
3563          Haste St at Telegraph Ave             37.866043   
...                                 ...                   ...   
505890        Fulton St at Bancroft Way             37.867789   
506183               37th St at West St             37.826696   
507540           Downtown Berkeley BART             37.870348   
507688        Fulton St at Bancroft Way             37.867789   
508021  MacArthur Blvd at Telegraph Ave             37.826286   

        end_station_longitude  bike_id   user_type  
1308              -122.267764     3474  Subscriber  
2915              -122.261351      311    Customer  
2917              -122.261351     2673    Customer  
3006              -122.253569     2129  Subscriber  
3563              -122.258804     3234    Customer  
...                       ...      ...         ...  
505890            -122.265896     1224  Subscriber  
506183            -122.271795     1321  Subscriber  
507540            -122.267764     1305  Subscriber  
507688            -122.265896     1219    Customer  
508021            -122.265100     1363  Subscriber  

[739 rows x 13 columns]

In [39]:
d

duration_sec                start_time                  end_time  \
1308             140  2017-12-31 09:45:30.2010  2017-12-31 09:47:50.4040   
2915            1006  2017-12-29 21:45:41.2250  2017-12-29 22:02:27.6540   
2917            1015  2017-12-29 21:45:11.1610  2017-12-29 22:02:06.3310   
3006             664  2017-12-29 19:45:00.8570  2017-12-29 19:56:05.6680   
3563            1401  2017-12-29 15:09:46.6360  2017-12-29 15:33:07.8160   
...              ...                       ...                       ...   
505890          2351  2017-07-12 20:59:56.7640  2017-07-12 21:39:08.4470   
506183          1279  2017-07-12 17:52:44.1640  2017-07-12 18:14:03.2870   
507540           403  2017-07-11 19:33:26.7920  2017-07-11 19:40:10.6640   
507688          8210  2017-07-11 16:10:08.2800  2017-07-11 18:26:58.4900   
508021          2031  2017-07-11 16:14:52.1460  2017-07-11 16:48:44.1150   

        start_station_id         start_station_name  start_station_latitude  \
1308                 247  Fulton St at Bancroft Way               37.867789   
2915                 247  Fulton St at Bancroft Way               37.867789   
2917                 247  Fulton St at Bancroft Way               37.867789   
3006                 247  Fulton St at Bancroft Way               37.867789   
3563                 247  Fulton St at Bancroft Way               37.867789   
...                  ...                        ...                     ...   
505890               247  Fulton St at Bancroft Way               37.867789   
506183               247  Fulton St at Bancroft Way               37.867789   
507540               247  Fulton St at Bancroft Way               37.867789   
507688               247  Fulton St at Bancroft Way               37.867789   
508021               247  Fulton St at Bancroft Way               37.867789   

        start_station_longitude  end_station_id  \
1308                -122.265896             245   
2915                -122.265896             170   
2917                -122.265896             170   
3006                -122.265896             267   
3563                -122.265896             240   
...                         ...             ...   
505890              -122.265896             247   
506183              -122.265896             192   
507540              -122.265896             245   
507688              -122.265896             247   
508021              -122.265896             177   

                       end_station_name  end_station_latitude  \
1308             Downtown Berkeley BART             37.870348   
2915           Telegraph Ave at 58th St             37.844493   
2917           Telegraph Ave at 58th St             37.844493   
3006            Derby St at College Ave             37.861804   
3563          Haste St at Telegraph Ave             37.866043   
...                                 ...                   ...   
505890        Fulton St at Bancroft Way             37.867789   
506183               37th St at West St             37.826696   
507540           Downtown Berkeley BART             37.870348   
507688        Fulton St at Bancroft Way             37.867789   
508021  MacArthur Blvd at Telegraph Ave             37.826286   

        end_station_longitude  bike_id   user_type  
1308              -122.267764     3474  Subscriber  
2915              -122.261351      311    Customer  
2917              -122.261351     2673    Customer  
3006              -122.253569     2129  Subscriber  
3563              -122.258804     3234    Customer  
...                       ...      ...         ...  
505890            -122.265896     1224  Subscriber  
506183            -122.271795     1321  Subscriber  
507540            -122.267764     1305  Subscriber  
507688            -122.265896     1219    Customer  
508021            -122.265100     1363  Subscriber  

[739 rows x 13 columns]

### which bike is used with less duration?

In [40]:
min_duration=df1[df1["duration_sec"]==df1["duration_sec"].min()]
min_duration

duration_sec                start_time                  end_time  \
8162              61  2017-12-27 08:40:13.2170  2017-12-27 08:41:14.9730   
14236             61  2017-12-22 07:20:37.1320  2017-12-22 07:21:38.6260   
16309             61  2017-12-21 10:23:30.0270  2017-12-21 10:24:31.6890   
17025             61  2017-12-21 08:09:41.4820  2017-12-21 08:10:42.7770   
21866             61  2017-12-19 17:31:47.6230  2017-12-19 17:32:49.4160   
23470             61  2017-12-19 09:35:32.4040  2017-12-19 09:36:34.0630   
44994             61  2017-12-12 20:25:24.0090  2017-12-12 20:26:25.1610   
64776             61  2017-12-07 08:53:02.6220  2017-12-07 08:54:04.3670   
65819             61  2017-12-06 21:18:33.9410  2017-12-06 21:19:35.0940   
69333             61  2017-12-06 08:37:23.8790  2017-12-06 08:38:25.6420   
80408             61  2017-12-02 22:14:25.6620  2017-12-02 22:15:26.7440   
84393             61  2017-12-01 14:24:58.7900  2017-12-01 14:26:00.0190   
94610             61  2017-11-29 08:14:22.7120  2017-11-29 08:15:23.8830   
102421            61  2017-11-27 08:44:18.6410  2017-11-27 08:45:19.9510   
114535            61  2017-11-21 08:23:34.9920  2017-11-21 08:24:36.1750   
119395            61  2017-11-19 18:15:27.2320  2017-11-19 18:16:28.5060   
133870            61  2017-11-14 17:11:43.3300  2017-11-14 17:12:44.4950   
165924            61  2017-11-05 20:13:06.2850  2017-11-05 20:14:07.3780   
186841            61  2017-10-30 19:13:40.4280  2017-10-30 19:14:41.6620   
195131            61  2017-10-27 18:16:21.2600  2017-10-27 18:17:22.6160   
198700            61  2017-10-26 21:32:43.7300  2017-10-26 21:33:45.0450   
204284            61  2017-10-25 18:11:38.8510  2017-10-25 18:12:40.7720   
219367            61  2017-10-21 17:16:15.5330  2017-10-21 17:17:16.5750   
219427            61  2017-10-21 16:56:45.3980  2017-10-21 16:57:47.2580   
219658            61  2017-10-21 16:02:05.0530  2017-10-21 16:03:06.6460   
227339            61  2017-10-19 09:48:20.5930  2017-10-19 09:49:22.4720   
246408            61  2017-10-13 15:19:13.5250  2017-10-13 15:20:15.4700   
248348            61  2017-10-12 21:57:40.8500  2017-10-12 21:58:42.4220   
260599            61  2017-10-09 17:54:17.8270  2017-10-09 17:55:19.0750   
264159            61  2017-10-08 15:21:55.0460  2017-10-08 15:22:56.9130   
270182            61  2017-10-06 12:51:19.3160  2017-10-06 12:52:20.3490   
276445            61  2017-10-04 20:05:18.0140  2017-10-04 20:06:19.7200   
292732            61  2017-09-30 12:48:57.8090  2017-09-30 12:49:59.2720   
293667            61  2017-09-29 20:48:20.2160  2017-09-29 20:49:21.3050   
312919            61  2017-09-25 07:51:47.8540  2017-09-25 07:52:49.6370   
313189            61  2017-09-25 00:21:24.0780  2017-09-25 00:22:25.8030   
316476            61  2017-09-23 15:22:17.5850  2017-09-23 15:23:19.3400   
332554            61  2017-09-19 09:21:32.3070  2017-09-19 09:22:33.3960   
346102            61  2017-09-14 21:27:39.8500  2017-09-14 21:28:41.5450   
376309            61  2017-09-06 08:34:43.2670  2017-09-06 08:35:44.8670   
381108            61  2017-09-04 17:56:04.2930  2017-09-04 17:57:05.4110   
383732            61  2017-09-03 14:08:28.9530  2017-09-03 14:09:30.7390   
383812            61  2017-09-03 13:39:22.8010  2017-09-03 13:40:23.8130   
386883            61  2017-09-01 19:38:27.3270  2017-09-01 19:39:29.0700   
395252            61  2017-08-30 13:51:38.2910  2017-08-30 13:52:39.3510   
395719            61  2017-08-30 10:20:40.8670  2017-08-30 10:21:42.1180   
396097            61  2017-08-30 09:04:16.3740  2017-08-30 09:05:18.0480   
401675            61  2017-08-28 15:38:18.4460  2017-08-28 15:39:19.6070   
406132            61  2017-08-26 16:50:40.1780  2017-08-26 16:51:42.0870   
420504            61  2017-08-22 00:40:36.2750  2017-08-22 00:41:37.9200   
420942            61  2017-08-21 18:52:53.6350  2017-08-21 18:53:55.2830   
445176            61  2017-08-12 14:49:18.1540  2017-08-

so, there are lot of bikes with minimum duration.

we can find total number of bikes with minimum duration by using below code

In [41]:
len(min_duration)

58

there are 68 bikes with same minimum duration

in this 68 bikes, we can find how many Customer and Subscriber type users by using below code

In [42]:
print("Total number of Customer type users whose duration time is minimum : ",len(min_duration[min_duration["user_type"]=="Customer"]))

print("Total number of Customer type users whose duration time is minimum : ",len(min_duration[min_duration["user_type"]=="Subscriber"]))


Total number of Customer type users whose duration time is minimum :  8
Total number of Customer type users whose duration time is minimum :  50


so, there are 8 Custermers and 50 subscribers with mimimum duration speed

### what is the bike id which has the maximum duration time?

In [43]:
print("bike id which has maximum duration time is : ",df1[df1["duration_sec"]==df1["duration_sec"].max()]["bike_id"])

bike id which has maximum duration time is :  138862    2231
Name: bike_id, dtype: int64


In [44]:
df1.columns

Index(['duration_sec', 'start_time', 'end_time', 'start_station_id',
       'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'end_station_id', 'end_station_name',
       'end_station_latitude', 'end_station_longitude', 'bike_id',
       'user_type'],
      dtype='object')

### Change the duration into time minutes from seconds?

In [45]:
df1['duration_min'] = df1['duration_sec'] / 60  
df1

duration_sec                start_time                  end_time  \
0              80110  2017-12-31 16:57:39.6540  2018-01-01 15:12:50.2450   
1              78800  2017-12-31 15:56:34.8420  2018-01-01 13:49:55.6170   
2              45768  2017-12-31 22:45:48.4110  2018-01-01 11:28:36.8830   
3              62172  2017-12-31 17:31:10.6360  2018-01-01 10:47:23.5310   
4              43603  2017-12-31 14:23:14.0010  2018-01-01 02:29:57.5710   
...              ...                       ...                       ...   
519695           435  2017-06-28 10:00:54.5280  2017-06-28 10:08:10.4380   
519696           431  2017-06-28 09:56:39.6310  2017-06-28 10:03:51.0900   
519697           424  2017-06-28 09:47:36.3470  2017-06-28 09:54:41.1870   
519698           366  2017-06-28 09:47:41.6640  2017-06-28 09:53:47.7150   
519699           188  2017-06-28 09:49:46.3770  2017-06-28 09:52:55.3380   

        start_station_id                                 start_station_name  \
0                     74                              Laguna St at Hayes St   
1                    284  Yerba Buena Center for the Arts (Howard St at ...   
2                    245                             Downtown Berkeley BART   
3                     60                               8th St at Ringold St   
4                    239                      Bancroft Way at Telegraph Ave   
...                  ...                                                ...   
519695                81                                 Berry St at 4th St   
519696                66                              3rd St at Townsend St   
519697                21   Montgomery St BART Station (Market St at 2nd St)   
519698                58                               Market St at 10th St   
519699                25                                Howard St at 2nd St   

        start_station_latitude  start_station_longitude  end_station_id  \
0                    37.776435              -122.426244              43   
1                    37.784872              -122.400876              96   
2                    37.870348              -122.267764             245   
3                    37.774520              -122.409449               5   
4                    37.868813              -122.258764             247   
...                        ...                      ...             ...   
519695               37.775880              -122.393170              45   
519696               37.778742              -122.392741             321   
519697               37.789625              -122.400811              48   
519698               37.776619              -122.417385              59   
519699               37.787522              -122.397405              48   

                                         end_station_name  \
0       San Francisco Public Library (Grove St at Hyde...   
1                                   Dolores St at 15th St   
2                                  Downtown Berkeley BART   
3            Powell St BART Station (Market St at 5th St)   
4                               Fulton St at Bancroft Way   
...                                                   ...   
519695                                5th St at Howard St   
519696                                      5th at Folsom   
519697                                2nd St at S Park St   
519698                        S Van Ness Ave at Market St   
519699                                2nd St at S Park St   

        end_station_latitude  end_station_longitude  bike_id   user_type  \
0                  37.778768            -122.415929       96    Customer   
1                  37.766210            -122.426614       88    Customer   
2                  37.870348            -122.267764     1094    Customer   
3                  37.783899            -122.408445     2831    Customer   
4                  37.867789            -122.265896     3167  Subscriber   
...                      ...                    ...     

### Change the duration into time hours from minutes?

In [46]:
df1['duration_hour'] = df1['duration_min'] / 60

In [29]:
df1

duration_sec                start_time                  end_time  \
0              80110  2017-12-31 16:57:39.6540  2018-01-01 15:12:50.2450   
1              78800  2017-12-31 15:56:34.8420  2018-01-01 13:49:55.6170   
2              45768  2017-12-31 22:45:48.4110  2018-01-01 11:28:36.8830   
3              62172  2017-12-31 17:31:10.6360  2018-01-01 10:47:23.5310   
4              43603  2017-12-31 14:23:14.0010  2018-01-01 02:29:57.5710   
...              ...                       ...                       ...   
519695           435  2017-06-28 10:00:54.5280  2017-06-28 10:08:10.4380   
519696           431  2017-06-28 09:56:39.6310  2017-06-28 10:03:51.0900   
519697           424  2017-06-28 09:47:36.3470  2017-06-28 09:54:41.1870   
519698           366  2017-06-28 09:47:41.6640  2017-06-28 09:53:47.7150   
519699           188  2017-06-28 09:49:46.3770  2017-06-28 09:52:55.3380   

        start_station_id                                 start_station_name  \
0                     74                              Laguna St at Hayes St   
1                    284  Yerba Buena Center for the Arts (Howard St at ...   
2                    245                             Downtown Berkeley BART   
3                     60                               8th St at Ringold St   
4                    239                      Bancroft Way at Telegraph Ave   
...                  ...                                                ...   
519695                81                                 Berry St at 4th St   
519696                66                              3rd St at Townsend St   
519697                21   Montgomery St BART Station (Market St at 2nd St)   
519698                58                               Market St at 10th St   
519699                25                                Howard St at 2nd St   

        start_station_latitude  start_station_longitude  end_station_id  \
0                    37.776435              -122.426244              43   
1                    37.784872              -122.400876              96   
2                    37.870348              -122.267764             245   
3                    37.774520              -122.409449               5   
4                    37.868813              -122.258764             247   
...                        ...                      ...             ...   
519695               37.775880              -122.393170              45   
519696               37.778742              -122.392741             321   
519697               37.789625              -122.400811              48   
519698               37.776619              -122.417385              59   
519699               37.787522              -122.397405              48   

                                         end_station_name  \
0       San Francisco Public Library (Grove St at Hyde...   
1                                   Dolores St at 15th St   
2                                  Downtown Berkeley BART   
3            Powell St BART Station (Market St at 5th St)   
4                               Fulton St at Bancroft Way   
...                                                   ...   
519695                                5th St at Howard St   
519696                                      5th at Folsom   
519697                                2nd St at S Park St   
519698                        S Van Ness Ave at Market St   
519699                                2nd St at S Park St   

        end_station_latitude  end_station_longitude  bike_id   user_type  \
0                  37.778768            -122.415929       96    Customer   
1                  37.766210            -122.426614       88    Customer   
2                  37.870348            -122.267764     1094    Customer   
3                  37.783899            -122.408445     2831    Customer   
4                  37.867789            -122.265896     3167  Subscriber   
...                      ...                    ...     

In [47]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 519700 entries, 0 to 519699
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   duration_sec             519700 non-null  int64  
 1   start_time               519700 non-null  object 
 2   end_time                 519700 non-null  object 
 3   start_station_id         519700 non-null  int64  
 4   start_station_name       519700 non-null  object 
 5   start_station_latitude   519700 non-null  float64
 6   start_station_longitude  519700 non-null  float64
 7   end_station_id           519700 non-null  int64  
 8   end_station_name         519700 non-null  object 
 9   end_station_latitude     519700 non-null  float64
 10  end_station_longitude    519700 non-null  float64
 11  bike_id                  519700 non-null  int64  
 12  user_type                519700 non-null  object 
 13  duration_min             519700 non-null  float64
 14  dura

In [48]:
#df['modified_start_time'] = pd.to_datetime(df['start_time '], format = "%Y-%m-%d")

In [49]:
df1["duration_hour"]

0         22.252778
1         21.888889
2         12.713333
3         17.270000
4         12.111944
            ...    
519695     0.120833
519696     0.119722
519697     0.117778
519698     0.101667
519699     0.052222
Name: duration_hour, Length: 519700, dtype: float64

In [50]:
df1.start_time = pd.to_datetime(df1.start_time)
df1.end_time = pd.to_datetime(df1.end_time)

In [51]:
df1

duration_sec              start_time                end_time  \
0              80110 2017-12-31 16:57:39.654 2018-01-01 15:12:50.245   
1              78800 2017-12-31 15:56:34.842 2018-01-01 13:49:55.617   
2              45768 2017-12-31 22:45:48.411 2018-01-01 11:28:36.883   
3              62172 2017-12-31 17:31:10.636 2018-01-01 10:47:23.531   
4              43603 2017-12-31 14:23:14.001 2018-01-01 02:29:57.571   
...              ...                     ...                     ...   
519695           435 2017-06-28 10:00:54.528 2017-06-28 10:08:10.438   
519696           431 2017-06-28 09:56:39.631 2017-06-28 10:03:51.090   
519697           424 2017-06-28 09:47:36.347 2017-06-28 09:54:41.187   
519698           366 2017-06-28 09:47:41.664 2017-06-28 09:53:47.715   
519699           188 2017-06-28 09:49:46.377 2017-06-28 09:52:55.338   

        start_station_id                                 start_station_name  \
0                     74                              Laguna St at Hayes St   
1                    284  Yerba Buena Center for the Arts (Howard St at ...   
2                    245                             Downtown Berkeley BART   
3                     60                               8th St at Ringold St   
4                    239                      Bancroft Way at Telegraph Ave   
...                  ...                                                ...   
519695                81                                 Berry St at 4th St   
519696                66                              3rd St at Townsend St   
519697                21   Montgomery St BART Station (Market St at 2nd St)   
519698                58                               Market St at 10th St   
519699                25                                Howard St at 2nd St   

        start_station_latitude  start_station_longitude  end_station_id  \
0                    37.776435              -122.426244              43   
1                    37.784872              -122.400876              96   
2                    37.870348              -122.267764             245   
3                    37.774520              -122.409449               5   
4                    37.868813              -122.258764             247   
...                        ...                      ...             ...   
519695               37.775880              -122.393170              45   
519696               37.778742              -122.392741             321   
519697               37.789625              -122.400811              48   
519698               37.776619              -122.417385              59   
519699               37.787522              -122.397405              48   

                                         end_station_name  \
0       San Francisco Public Library (Grove St at Hyde...   
1                                   Dolores St at 15th St   
2                                  Downtown Berkeley BART   
3            Powell St BART Station (Market St at 5th St)   
4                               Fulton St at Bancroft Way   
...                                                   ...   
519695                                5th St at Howard St   
519696                                      5th at Folsom   
519697                                2nd St at S Park St   
519698                        S Van Ness Ave at Market St   
519699                                2nd St at S Park St   

        end_station_latitude  end_station_longitude  bike_id   user_type  \
0                  37.778768            -122.415929       96    Customer   
1                  37.766210            -122.426614       88    Customer   
2                  37.870348            -122.267764     1094    Customer   
3                  37.783899            -122.408445     2831    Customer   
4                  37.867789            -122.265896     3167  Subscriber   
...                      ...                    ...      ...         ...   
519695             37.781752

### Changing the change the data type for the column start_time and end_time to datetime type?

In [52]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 519700 entries, 0 to 519699
Data columns (total 15 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   duration_sec             519700 non-null  int64         
 1   start_time               519700 non-null  datetime64[ns]
 2   end_time                 519700 non-null  datetime64[ns]
 3   start_station_id         519700 non-null  int64         
 4   start_station_name       519700 non-null  object        
 5   start_station_latitude   519700 non-null  float64       
 6   start_station_longitude  519700 non-null  float64       
 7   end_station_id           519700 non-null  int64         
 8   end_station_name         519700 non-null  object        
 9   end_station_latitude     519700 non-null  float64       
 10  end_station_longitude    519700 non-null  float64       
 11  bike_id                  519700 non-null  int64         
 12  user_type       

### Creating the column month name,month number,day name,date and hour when started by using start_time column?

In [55]:
df1['start_time_month_name']=df1['start_time'].dt.strftime('%B')
df1['start_time_month_number']=df1['start_time'].dt.month.astype(int)
df1['start_time_weekday']=df1['start_time'].dt.strftime('%a')
df1['start_time_day']=df1['start_time'].dt.day.astype(int)
df1['start_time_hour']=df1['start_time'].dt.hour

In [56]:
df1

duration_sec              start_time                end_time  \
0              80110 2017-12-31 16:57:39.654 2018-01-01 15:12:50.245   
1              78800 2017-12-31 15:56:34.842 2018-01-01 13:49:55.617   
2              45768 2017-12-31 22:45:48.411 2018-01-01 11:28:36.883   
3              62172 2017-12-31 17:31:10.636 2018-01-01 10:47:23.531   
4              43603 2017-12-31 14:23:14.001 2018-01-01 02:29:57.571   
...              ...                     ...                     ...   
519695           435 2017-06-28 10:00:54.528 2017-06-28 10:08:10.438   
519696           431 2017-06-28 09:56:39.631 2017-06-28 10:03:51.090   
519697           424 2017-06-28 09:47:36.347 2017-06-28 09:54:41.187   
519698           366 2017-06-28 09:47:41.664 2017-06-28 09:53:47.715   
519699           188 2017-06-28 09:49:46.377 2017-06-28 09:52:55.338   

        start_station_id                                 start_station_name  \
0                     74                              Laguna St at Hayes St   
1                    284  Yerba Buena Center for the Arts (Howard St at ...   
2                    245                             Downtown Berkeley BART   
3                     60                               8th St at Ringold St   
4                    239                      Bancroft Way at Telegraph Ave   
...                  ...                                                ...   
519695                81                                 Berry St at 4th St   
519696                66                              3rd St at Townsend St   
519697                21   Montgomery St BART Station (Market St at 2nd St)   
519698                58                               Market St at 10th St   
519699                25                                Howard St at 2nd St   

        start_station_latitude  start_station_longitude  end_station_id  \
0                    37.776435              -122.426244              43   
1                    37.784872              -122.400876              96   
2                    37.870348              -122.267764             245   
3                    37.774520              -122.409449               5   
4                    37.868813              -122.258764             247   
...                        ...                      ...             ...   
519695               37.775880              -122.393170              45   
519696               37.778742              -122.392741             321   
519697               37.789625              -122.400811              48   
519698               37.776619              -122.417385              59   
519699               37.787522              -122.397405              48   

                                         end_station_name  \
0       San Francisco Public Library (Grove St at Hyde...   
1                                   Dolores St at 15th St   
2                                  Downtown Berkeley BART   
3            Powell St BART Station (Market St at 5th St)   
4                               Fulton St at Bancroft Way   
...                                                   ...   
519695                                5th St at Howard St   
519696                                      5th at Folsom   
519697                                2nd St at S Park St   
519698                        S Van Ness Ave at Market St   
519699                                2nd St at S Park St   

        end_station_latitude  end_station_longitude  bike_id   user_type  \
0                  37.778768            -122.415929       96    Customer   
1                  37.766210            -122.426614       88    Customer   
2                  37.870348            -122.267764     1094    Customer   
3                  37.783899            -122.408445     2831    Customer   
4                  37.867789            -122.265896     3167  Subscriber   
...                      ...                    ...      ...         ...   
519695             37.781752

### creating the column month name,month number,day name,date and hour when ended by using end_time column

In [57]:
df1['end_time_month_name']=df1['end_time'].dt.strftime('%B')
df1['end_time_month_number']=df1['end_time'].dt.month.astype(int)
df1['end_time_weekday']=df1['end_time'].dt.strftime('%a')
df1['end_time_day']=df1['end_time'].dt.day.astype(int)
df1['end_time_hour']=df1['end_time'].dt.hour

In [39]:
df1

duration_sec                start_time                  end_time  \
0              80110  2017-12-31 16:57:39.6540  2018-01-01 15:12:50.2450   
1              78800  2017-12-31 15:56:34.8420  2018-01-01 13:49:55.6170   
2              45768  2017-12-31 22:45:48.4110  2018-01-01 11:28:36.8830   
3              62172  2017-12-31 17:31:10.6360  2018-01-01 10:47:23.5310   
4              43603  2017-12-31 14:23:14.0010  2018-01-01 02:29:57.5710   
...              ...                       ...                       ...   
519695           435  2017-06-28 10:00:54.5280  2017-06-28 10:08:10.4380   
519696           431  2017-06-28 09:56:39.6310  2017-06-28 10:03:51.0900   
519697           424  2017-06-28 09:47:36.3470  2017-06-28 09:54:41.1870   
519698           366  2017-06-28 09:47:41.6640  2017-06-28 09:53:47.7150   
519699           188  2017-06-28 09:49:46.3770  2017-06-28 09:52:55.3380   

        start_station_id                                 start_station_name  \
0                     74                              Laguna St at Hayes St   
1                    284  Yerba Buena Center for the Arts (Howard St at ...   
2                    245                             Downtown Berkeley BART   
3                     60                               8th St at Ringold St   
4                    239                      Bancroft Way at Telegraph Ave   
...                  ...                                                ...   
519695                81                                 Berry St at 4th St   
519696                66                              3rd St at Townsend St   
519697                21   Montgomery St BART Station (Market St at 2nd St)   
519698                58                               Market St at 10th St   
519699                25                                Howard St at 2nd St   

        start_station_latitude  start_station_longitude  end_station_id  \
0                    37.776435              -122.426244              43   
1                    37.784872              -122.400876              96   
2                    37.870348              -122.267764             245   
3                    37.774520              -122.409449               5   
4                    37.868813              -122.258764             247   
...                        ...                      ...             ...   
519695               37.775880              -122.393170              45   
519696               37.778742              -122.392741             321   
519697               37.789625              -122.400811              48   
519698               37.776619              -122.417385              59   
519699               37.787522              -122.397405              48   

                                         end_station_name  \
0       San Francisco Public Library (Grove St at Hyde...   
1                                   Dolores St at 15th St   
2                                  Downtown Berkeley BART   
3            Powell St BART Station (Market St at 5th St)   
4                               Fulton St at Bancroft Way   
...                                                   ...   
519695                                5th St at Howard St   
519696                                      5th at Folsom   
519697                                2nd St at S Park St   
519698                        S Van Ness Ave at Market St   
519699                                2nd St at S Park St   

        end_station_latitude  end_station_longitude  bike_id   user_type  \
0                  37.778768            -122.415929       96    Customer   
1                  37.766210            -122.426614       88    Customer   
2                  37.870348            -122.267764     1094    Customer   
3                  37.783899            -122.408445     2831    Customer   
4                  37.867789            -122.265896     3167  Subscriber   
...                      ...                    ...     

In [40]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 519700 entries, 0 to 519699
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   duration_sec             519700 non-null  int64  
 1   start_time               519700 non-null  object 
 2   end_time                 519700 non-null  object 
 3   start_station_id         519700 non-null  int64  
 4   start_station_name       519700 non-null  object 
 5   start_station_latitude   519700 non-null  float64
 6   start_station_longitude  519700 non-null  float64
 7   end_station_id           519700 non-null  int64  
 8   end_station_name         519700 non-null  object 
 9   end_station_latitude     519700 non-null  float64
 10  end_station_longitude    519700 non-null  float64
 11  bike_id                  519700 non-null  int64  
 12  user_type                519700 non-null  object 
 13  duration_min             519700 non-null  float64
dtypes: f

In [58]:
df1.columns

Index(['duration_sec', 'start_time', 'end_time', 'start_station_id',
       'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'end_station_id', 'end_station_name',
       'end_station_latitude', 'end_station_longitude', 'bike_id', 'user_type',
       'duration_min', 'duration_hour', 'start_time_month_name',
       'start_time_month_number', 'start_time_weekday', 'start_time_day',
       'start_time_hour', 'end_time_month_name', 'end_time_month_number',
       'end_time_weekday', 'end_time_day', 'end_time_hour'],
      dtype='object')

### Extract the data of the records which have the duration moretham 15 hours?

In [59]:
p=df1[df1["duration_sec"]>=15]
p

duration_sec              start_time                end_time  \
0              80110 2017-12-31 16:57:39.654 2018-01-01 15:12:50.245   
1              78800 2017-12-31 15:56:34.842 2018-01-01 13:49:55.617   
2              45768 2017-12-31 22:45:48.411 2018-01-01 11:28:36.883   
3              62172 2017-12-31 17:31:10.636 2018-01-01 10:47:23.531   
4              43603 2017-12-31 14:23:14.001 2018-01-01 02:29:57.571   
...              ...                     ...                     ...   
519695           435 2017-06-28 10:00:54.528 2017-06-28 10:08:10.438   
519696           431 2017-06-28 09:56:39.631 2017-06-28 10:03:51.090   
519697           424 2017-06-28 09:47:36.347 2017-06-28 09:54:41.187   
519698           366 2017-06-28 09:47:41.664 2017-06-28 09:53:47.715   
519699           188 2017-06-28 09:49:46.377 2017-06-28 09:52:55.338   

        start_station_id                                 start_station_name  \
0                     74                              Laguna St at Hayes St   
1                    284  Yerba Buena Center for the Arts (Howard St at ...   
2                    245                             Downtown Berkeley BART   
3                     60                               8th St at Ringold St   
4                    239                      Bancroft Way at Telegraph Ave   
...                  ...                                                ...   
519695                81                                 Berry St at 4th St   
519696                66                              3rd St at Townsend St   
519697                21   Montgomery St BART Station (Market St at 2nd St)   
519698                58                               Market St at 10th St   
519699                25                                Howard St at 2nd St   

        start_station_latitude  start_station_longitude  end_station_id  \
0                    37.776435              -122.426244              43   
1                    37.784872              -122.400876              96   
2                    37.870348              -122.267764             245   
3                    37.774520              -122.409449               5   
4                    37.868813              -122.258764             247   
...                        ...                      ...             ...   
519695               37.775880              -122.393170              45   
519696               37.778742              -122.392741             321   
519697               37.789625              -122.400811              48   
519698               37.776619              -122.417385              59   
519699               37.787522              -122.397405              48   

                                         end_station_name  \
0       San Francisco Public Library (Grove St at Hyde...   
1                                   Dolores St at 15th St   
2                                  Downtown Berkeley BART   
3            Powell St BART Station (Market St at 5th St)   
4                               Fulton St at Bancroft Way   
...                                                   ...   
519695                                5th St at Howard St   
519696                                      5th at Folsom   
519697                                2nd St at S Park St   
519698                        S Van Ness Ave at Market St   
519699                                2nd St at S Park St   

        end_station_latitude  ...  start_time_month_name  \
0                  37.778768  ...               December   
1                  37.766210  ...               December   
2                  37.870348  ...               December   
3                  37.783899  ...               December   
4                  37.867789  ...               December   
...                      ...  ...                    ...   
519695             37.781752  ...                   June   
519696             37.780146  ...                   June   
519697             3

### Extracting the user_type, starting and end station name for the bike id 1729?

In [60]:
df1.loc[1729,["user_type","start_station_name","end_station_name"]]

user_type                                Subscriber
start_station_name               Ashby BART Station
end_station_name      Telegraph Ave at Alcatraz Ave
Name: 1729, dtype: object

In [61]:
df1.columns

Index(['duration_sec', 'start_time', 'end_time', 'start_station_id',
       'start_station_name', 'start_station_latitude',
       'start_station_longitude', 'end_station_id', 'end_station_name',
       'end_station_latitude', 'end_station_longitude', 'bike_id', 'user_type',
       'duration_min', 'duration_hour', 'start_time_month_name',
       'start_time_month_number', 'start_time_weekday', 'start_time_day',
       'start_time_hour', 'end_time_month_name', 'end_time_month_number',
       'end_time_weekday', 'end_time_day', 'end_time_hour'],
      dtype='object')

### Between Customers and the Subscribers who are using the bike services?

In [62]:
df[df['user_type']=='Customer']

duration_sec                start_time                  end_time  \
0              80110  2017-12-31 16:57:39.6540  2018-01-01 15:12:50.2450   
1              78800  2017-12-31 15:56:34.8420  2018-01-01 13:49:55.6170   
2              45768  2017-12-31 22:45:48.4110  2018-01-01 11:28:36.8830   
3              62172  2017-12-31 17:31:10.6360  2018-01-01 10:47:23.5310   
5               9226  2017-12-31 22:51:00.9180  2018-01-01 01:24:47.1660   
...              ...                       ...                       ...   
519631          3819  2017-06-28 10:44:43.7170  2017-06-28 11:48:23.6360   
519636          5441  2017-06-28 10:11:30.6360  2017-06-28 11:42:12.5950   
519639          1562  2017-06-28 11:08:51.3220  2017-06-28 11:34:53.6830   
519643          1778  2017-06-28 10:54:52.4630  2017-06-28 11:24:30.5330   
519663          2349  2017-06-28 10:16:48.2310  2017-06-28 10:55:57.2800   

        start_station_id                                 start_station_name  \
0                     74                              Laguna St at Hayes St   
1                    284  Yerba Buena Center for the Arts (Howard St at ...   
2                    245                             Downtown Berkeley BART   
3                     60                               8th St at Ringold St   
5                     30     San Francisco Caltrain (Townsend St at 4th St)   
...                  ...                                                ...   
519631                48                                2nd St at S Park St   
519636                45                                5th St at Howard St   
519639                42      San Francisco City Hall (Polk St at Grove St)   
519643                15  San Francisco Ferry Building (Harry Bridges Pl...   
519663                28                       The Embarcadero at Bryant St   

        start_station_latitude  start_station_longitude  end_station_id  \
0                    37.776435              -122.426244              43   
1                    37.784872              -122.400876              96   
2                    37.870348              -122.267764             245   
3                    37.774520              -122.409449               5   
5                    37.776598              -122.395282              30   
...                        ...                      ...             ...   
519631               37.782411              -122.392706              48   
519636               37.781752              -122.405127              42   
519639               37.778650              -122.418230              23   
519643               37.795392              -122.394203              42   
519663               37.787168              -122.388098              42   

                                         end_station_name  \
0       San Francisco Public Library (Grove St at Hyde...   
1                                   Dolores St at 15th St   
2                                  Downtown Berkeley BART   
3            Powell St BART Station (Market St at 5th St)   
5          San Francisco Caltrain (Townsend St at 4th St)   
...                                                   ...   
519631                                2nd St at S Park St   
519636      San Francisco City Hall (Polk St at Grove St)   
519639                      The Embarcadero at Steuart St   
519643      San Francisco City Hall (Polk St at Grove St)   
519663      San Francisco City Hall (Polk St at Grove St)   

        end_station_latitude  end_station_longitude  bike_id user_type  
0                  37.778768            -122.415929       96  Customer  
1                  37.766210            -122.426614       88  Customer  
2                  37.870348            -122.267764     1094  Customer  
3                  37.783899            -122.408445     2831  Customer  
5                  37.776598            -122.395282     1487  Customer  
...                      ...                    ...      ...       ...  
5

duration_sec                start_time                  end_time  \
0              80110  2017-12-31 16:57:39.6540  2018-01-01 15:12:50.2450   
1              78800  2017-12-31 15:56:34.8420  2018-01-01 13:49:55.6170   
2              45768  2017-12-31 22:45:48.4110  2018-01-01 11:28:36.8830   
3              62172  2017-12-31 17:31:10.6360  2018-01-01 10:47:23.5310   
4              43603  2017-12-31 14:23:14.0010  2018-01-01 02:29:57.5710   
...              ...                       ...                       ...   
519695           435  2017-06-28 10:00:54.5280  2017-06-28 10:08:10.4380   
519696           431  2017-06-28 09:56:39.6310  2017-06-28 10:03:51.0900   
519697           424  2017-06-28 09:47:36.3470  2017-06-28 09:54:41.1870   
519698           366  2017-06-28 09:47:41.6640  2017-06-28 09:53:47.7150   
519699           188  2017-06-28 09:49:46.3770  2017-06-28 09:52:55.3380   

        start_station_id                                 start_station_name  \
0                     74                              Laguna St at Hayes St   
1                    284  Yerba Buena Center for the Arts (Howard St at ...   
2                    245                             Downtown Berkeley BART   
3                     60                               8th St at Ringold St   
4                    239                      Bancroft Way at Telegraph Ave   
...                  ...                                                ...   
519695                81                                 Berry St at 4th St   
519696                66                              3rd St at Townsend St   
519697                21   Montgomery St BART Station (Market St at 2nd St)   
519698                58                               Market St at 10th St   
519699                25                                Howard St at 2nd St   

        start_station_latitude  start_station_longitude  end_station_id  \
0                    37.776435              -122.426244              43   
1                    37.784872              -122.400876              96   
2                    37.870348              -122.267764             245   
3                    37.774520              -122.409449               5   
4                    37.868813              -122.258764             247   
...                        ...                      ...             ...   
519695               37.775880              -122.393170              45   
519696               37.778742              -122.392741             321   
519697               37.789625              -122.400811              48   
519698               37.776619              -122.417385              59   
519699               37.787522              -122.397405              48   

                                         end_station_name  \
0       San Francisco Public Library (Grove St at Hyde...   
1                                   Dolores St at 15th St   
2                                  Downtown Berkeley BART   
3            Powell St BART Station (Market St at 5th St)   
4                               Fulton St at Bancroft Way   
...                                                   ...   
519695                                5th St at Howard St   
519696                                      5th at Folsom   
519697                                2nd St at S Park St   
519698                        S Van Ness Ave at Market St   
519699                                2nd St at S Park St   

        end_station_latitude  end_station_longitude  bike_id   user_type  
0                  37.778768            -122.415929       96    Customer  
1                  37.766210            -122.426614       88    Customer  
2                  37.870348            -122.267764     1094    Customer  
3                  37.783899            -122.408445     2831    Customer  
4                  37.867789            -122.265896     3167  Subscriber  
...                      ...                    ...      ...  